early_stopping/simulator.py

```diff
@@ -360,7 +368,9 @@ def addCallback(self, callback: SimulationCallback) -> None:
             raise ValueError(f"Callbacks must be of type {type(SimulationCallback)}!")
 
         if callback.has_save_artifacts:
-            callback.path = self.output_dir
+            callback.path = os.path.join(self.output_dir, self.callback_dir_name)
+            if not os.path.exists(callback.path):
+                os.mkdir(callback.path)
```

Comment:
callbacks should be the owner of their own logic. Make a method in callbacks class that does this and is called by the simulator.